File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
       Map<Integer,Integer> presum = new HashMap<>();
       int count = 0;
       int currsum = 0;
       for(int i = 0;i < nums.length;i++){
        currsum += nums[i];
        if(currsum == k) count += 1;
        if(presum.containsKey(currsum - k)){
            count += presum.get(currsum - k);
        }
        presum.put(currsum,presum.getOrDefault(currsum,0) + 1);
       }
       return count;
       
    }
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        map<int,int> map;
        int count = 0;
        int sum = 0;
        map[0] = 1;
        for(int i = 0;i < nums.size();i++){
            sum += nums[i];
            int remove = sum - k;
            count += map[remove];
            map[sum]++;
        }
        return count;     
    }
};